### DISCUSSION TITLE
Questions about Inputs of Unreal Engine & JSBSim

### DISCUSSION THREAD

**Discussion by hesetone** [JSBSim-Team/jsbsim#771] [Q&A]
Hello, I am learning about JSBSim&UE5, but there is something puzzles me. Now, the `Flight Commands` in [jsbsim/UnrealEngine/README.md](https://github.com/JSBSim-Team/jsbsim/blob/master/UnrealEngine/README.md) are not suitable for my project, I want to formulate commands by myself and send them to JSBSim via UDP socket. As we can see, the contents of 737.xml(L896-L07) indicates properties that will be sent to JSBSim:
https://github.com/JSBSim-Team/jsbsim/blob/8fb77410bb8457d6bfe517e1df34660dc0df4373/aircraft/737/737.xml#L896-L907
so, the point puzzles me shows  as below:
1, I  don't want to change any other states of the 737 plane, but just set elevator to -0.3. Should i formulate my command as `0.0,-0.3,0.0,0.0,0.0,0`? 
2, what the meaning of `+ 1 for the time stamp` in 737.xml(line 899)? Does it means that I have to attach a time stamp at the header of socket contents? I have checked the rules of how socket contents will be interpreted in FGUDPInputSocket.cpp. it shows as below: https://github.com/JSBSim-Team/jsbsim/blob/8fb77410bb8457d6bfe517e1df34660dc0df4373/src/input_output/FGUDPInputSocket.cpp#L96-L129
how can i insert the time stamp into socket contents? the real world time is not defined by user, how can i insert the correct time into socket contents?(the description of question 2 maybe not clear enough, i will appreciate if you can help me figure it out.)

**Comment by bcoconni** [comment]
Hi @hesetone,

Regarding the first topic: I'd suggest sending the same command values than previously for all the parameters that you do not plan to change.

For example, say you sent `0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.0` at the previous iteration (see below about why there are seven values) and you now want to  change the elevator position to `-0.3` while keeping all the other parameters unchanged then you must send `1.0, 0.1, -0.3, 0.3, 0.4, 0.5, 0.0` (the time step is incremented and the elevator command has been replaced by `-0.3`).

The answer to the second topic is in the code below
https://github.com/JSBSim-Team/jsbsim/blob/8fb77410bb8457d6bfe517e1df34660dc0df4373/src/input_output/FGUDPInputSocket.cpp#L109-L123

You need to prepend the data you send by a time stamp and JSBSim will use it to make sure that the current data packet is newer than the last data packet that it has processed. The time stamp does not need to match the elapsed simulated time nor real time, its purpose is just to discard older data packets because I believe that the UDP format does not guarantee the packets to be received in the same order than they were sent.

@dpculp is the author of the `QTJSBSIM` format so he might correct me if I made incorrect statements.

**Reply by hesetone** [reply]
>  The time stamp does not need to match the elapsed simulated time nor real time, its purpose is just to discard older data packets.
Thank you so much, the description above would be of great help to me. But there is another question, I ran `UEReferenceAPP` in VS2019, but  i found the port `5139` was not used, when i run the UEReferenceAPP.sln, list of UDP services shows as below:
```
 > NETSTAT.EXE -ano
  UDP    0.0.0.0:53             *:*                                    3780
  UDP    0.0.0.0:5050           *:*                                    1580
  UDP    0.0.0.0:5353           *:*                                    23928
  UDP    0.0.0.0:5353           *:*                                    2184
  UDP    0.0.0.0:5353           *:*                                    23928
  UDP    0.0.0.0:5353           *:*                                    2184
  UDP    0.0.0.0:5353           *:*                                    3852
  UDP    0.0.0.0:5353           *:*                                    23928
  UDP    0.0.0.0:5353           *:*                                    2184
  UDP    0.0.0.0:5353           *:*                                    23928
  UDP    0.0.0.0:5353           *:*                                    2184
  UDP    0.0.0.0:5355           *:*                                    3852
  UDP    0.0.0.0:6666           *:*                                    29000
  UDP    0.0.0.0:6666           *:*                                    31556
  UDP    0.0.0.0:52213          *:*                                    31556
  UDP    0.0.0.0:56241          *:*                                    29624
  UDP    0.0.0.0:57811          *:*                                    29000
  UDP    0.0.0.0:60211          *:*                                    3780
  UDP    0.0.0.0:60212          *:*                                    3780
  UDP    127.0.0.1:1900         *:*                                    15204
  UDP    127.0.0.1:51229        *:*                                    5288
  UDP    127.0.0.1:63581        *:*                                    15204
  UDP    172.29.176.1:137       *:*                                    4
  UDP    172.29.176.1:138       *:*                                    4
  UDP    172.29.176.1:1900      *:*                                    15204
  UDP    172.29.176.1:63579     *:*                                    15204
  UDP    192.168.8.15:137       *:*                                    4
  UDP    192.168.8.15:138       *:*                                    4
  UDP    192.168.8.15:1900      *:*                                    15204
  UDP    192.168.8.15:63580     *:*                                    15204
 ......
```
If I run JSBSim as a independent program(named `JSBSimMain`) on Mac OS, It shows that the port `5139` is used:
```
> lsof -i:5139
COMMAND      PID       USER     FD    TYPE                      DEVICE  SIZE/OFF MODE NAME 
JSBSimMain 52158 hesetone       4u    IPv4  0xf1706bfd6c96469            0t0    UDP *:5139
```
as we can see, if we have `<input porrt="5139" type="QTJSBSIM" rate="20">` in 737.xml, function `FGUDPInputSocket::FGUDPInputSocket(FGFDMExec* fdmex)` will be called, then we will find the port `5139` is used. but I cannot get the same result while running JSBSim in UEReferenceAPP. Is there some key points i have missed?

**Reply by seanmcleod** [reply]
Are you sure that you have configured the `UEReferenceApp` to use the 737 model and it isn't defaulting to loading some other model? I don't think any of the other aircraft models in the JSBSim repo by default have a `QTJSBSIM` input section.

**Comment by seanmcleod** [comment]
> The time stamp does not need to match the elapsed simulated time nor real time, its purpose is just to discard older data packets

Correct, so you can view the **timestamp** as effectively a sequence counter.

> I believe that the UDP format does not guarantee the packets to be received in the same order than they were sent.

The UDP protocol, unlike the TCP protocol, first of all doesn't guarantee delivery, i.e. there is no automatic re-transmission for missing packets etc. by the OS, and secondly, again unlike TCP doesn't guarantee they will be received in the same order that they were transmitted.

Although realistically when using UDP between 2 processes on the same PC or even on a LAN as opposed to over the internet across multiple routers etc. you won't typically see non-delivery or out of order delivery.

But the code receiving the UDP packets in JSBSim has to take the possibility into account, so it quietly drops/ignores an update if it is older that the last one if received.

**Reply by bcoconni** [reply]
Thanks for the details @seanmcleod 👍
> Correct, so you can view the **timestamp** as effectively a sequence counter.

For my information, is the correct spelling "timestamp" (one word) or "time stamp" (two words) ? I have mimicked the spelling from the code (I.e. "time stamp") but I really can't tell which one is correct in this context (may be both ?)

**Comment by seanmcleod** [comment]
Hmm, although English is my first language, I'm not 100% sure 😉 

Pretty sure both are in use. I made it one word to emphasize the variable name in the code more than anything else.

Merriam Webster seems to lump both of them together.

https://www.merriam-webster.com/dictionary/timestamp

**Reply by bcoconni** [reply]
Ha OK. Thanks ! 😄

**Comment by MariaALJ** [comment]
Hello, I have a question from the same subject: the input of unreal engine. Can someone help me about the normalization please: if the elevator command is at 0.3, what does that mean in deg or rad? Thank you in advance!

**Reply by seanmcleod** [reply]
You need to look at the specific aircraft's FDM file to figure out the mapping between normalized positions and angular positions.

https://github.com/JSBSim-Team/jsbsim/blob/c4970ed343c9e4b49532bd168a5e3626e8466e6e/aircraft/737/737.xml#L189-L224

The control surface's angular position will then be used in the FDM to calculate forces and moments.

https://github.com/JSBSim-Team/jsbsim/blob/c4970ed343c9e4b49532bd168a5e3626e8466e6e/aircraft/737/737.xml#L622-L630

https://github.com/JSBSim-Team/jsbsim/blob/c4970ed343c9e4b49532bd168a5e3626e8466e6e/aircraft/737/737.xml#L681-L689

https://github.com/JSBSim-Team/jsbsim/blob/c4970ed343c9e4b49532bd168a5e3626e8466e6e/aircraft/737/737.xml#L770-L785

**Reply by MariaALJ** [reply]
Thank you very much!

**Comment by MariaALJ** [comment]
I have another question related to unreal engine please: How can i change the aircraft in the UEreference application? Are all the aircraft files available adapatble to the UEReference App?

**Reply by MariaALJ** [reply]
I've read issue #625 that is now closed but it didn't work for me:

AlbanBERGERET-Epic > Yes, the Unreal Engine plugin is made in such a way that when you change the aircraft name property on the jsbsimcomponent, it will load the corresponding jsbsim XML file. If you need another aircraft, or edit an existing one, you have to edit the model the jsbsim way in the aircraft database.

In the JSBSimMovement Component in details>Model>Aircraft Model i'm typing C130.xml (or C130) for exemple and nothing is changing. Thanks again in advance for your help!

**Reply by gallonmate** [reply]
In UE5, go to the BP airliner in the level editor, then go to it's JSBSim component. And type in the aircraft name as seen below.
![image](https://user-images.githubusercontent.com/28702753/204322912-3e1704e6-5366-4484-bd9c-fcb71ae769b7.png)

**Reply by MariaALJ** [reply]
Thank you for your response. This is exactly what I'm doing but it's not working: When I change the aircarft model it either doesn’t change anything or the default 787-8 (that keeps on apearing) starts having bizarre behavior like turning around it self while on ground. This is a glimpse of the error I’m having. Thanks in adavance for your help again!!


Blueprint Runtime Error: “Attempted to access index 1 from array EngineStates of length 1!”. Node: Set RightEnginePlayRate Graph: EventGraph Function: Execute Ubergraph ABP Commercial Plane Blueprint: ABP_CommercialPlane
Blueprint Runtime Error: “Accessed None reading structure N1”. Node: Set RightEnginePlayRate Graph: EventGraph Function: Execute Ubergraph ABP Commercial Plane Blueprint: ABP_CommercialPlane
Blueprint Runtime Error: “Attempted to access index 1 from array EngineStates of length 1!”. Node: Set RightEnginePlayRate Graph: EventGraph Function: Execute Ubergraph ABP Commercial Plane Blueprint: ABP_CommercialPlane
Blueprint Runtime Error: “Accessed None reading structure N1”. Node: Set RightEnginePlayRate Graph: EventGraph Function: Execute Ubergraph ABP Commercial Plane Blueprint: ABP_CommercialPlane
Blueprint Runtime Error: “Attempted to access index 1 from array EngineStates of length 1!”. Node: Set RightEnginePlayRate Graph: EventGraph Function: Execute Ubergraph ABP Commercial Plane Blueprint: ABP_CommercialPlane
Blueprint Runtime Error: “Accessed None reading structure N1”. Node: Set RightEnginePlayRate Graph: EventGraph Function: Execute Ubergraph ABP Commercial Plane Blueprint: ABP_CommercialPlane
Blueprint Runtime Error: “Attempted to access index 1 from array EngineStates of length 1!”. Node: Set Scalar Parameter Value Graph: EventGraph Function: Execute Ubergraph BP Airliner Blueprint: BP_Airliner
Blueprint Runtime Error: “Accessed None reading structure N1”. Node: Set Scalar Parameter Value Graph: EventGraph Function: Execute Ubergraph BP Airliner Blueprint: BP_Airliner
Blueprint Runtime Error: “Attempted to access index 1 from array EngineStates of length 1!”. Node: Set RightEnginePlayRate Graph: EventGraph Function: Execute Ubergraph ABP Commercial Plane Blueprint: ABP_CommercialPlane
Blueprint Runtime Error: “Accessed None reading structure N1”. Node: Set RightEnginePlayRate Graph: EventGraph Function: Execute Ubergraph ABP Commercial Plane Blueprint: ABP_CommercialPlane
Blueprint Runtime Error: “Attempted to access index 1 from array EngineStates of length 1!”. Node: Set RightEnginePlayRate Graph: EventGraph Function: Execute Ubergraph ABP Commercial Plane Blueprint: ABP_CommercialPlane
Blueprint Runtime Error: “Accessed None reading structure N1”. Node: Set RightEnginePlayRate Graph: EventGraph Function: Execute Ubergraph ABP Commercial Plane Blueprint: ABP_CommercialPlane
Blueprint Runtime Error: “Attempted to access index 1 from array EngineStates of length 1!”. Node: Set RightEnginePlayRate Graph: EventGraph Function: Execute Ubergraph ABP Commercial Plane Blueprint: ABP_CommercialPlane
Blueprint Runtime Error: “Accessed None reading structure N1”. Node: Set RightEnginePlayRate Graph: EventGraph Function: Execute Ubergraph ABP Commercial Plane Blueprint: ABP_CommercialPlane
Blueprint Runtime Error: “Attempted to access index 1 from array EngineStates of length 1!”. Node: Set Scalar Parameter Value Graph: EventGraph Function: Execute Ubergraph BP Airliner Blueprint: BP_Airliner

**Comment by seanmcleod** [comment]
Looks like you're switching from a 2-engine specific jet engine to possibly a 1-engine piston aircraft? My guess is if you try switching to say the 737 model you wouldn't see these errors.

I'm not really familiar with the UE plug-in, but I'm guessing there may be a separate single-engine piston blueprint option?

**Reply by MariaALJ** [reply]
Indeed with A320 and 737, I don't have errors but when I press play I still see the 787-8....

**Reply by MariaALJ** [reply]
Concerning the engines, I'm interested in 4 aircarft models : C130, f22, f450 and a helicopter. What can I do on the engine level to support these aircraft models?

**Comment by gallonmate** [comment]
@MariaALJ Please create a new discussion and provide more details of exactly what you're doing. We all thought you were the original author of this discussion and we are making assumptions based on the original author's details....

**Comment by seanmcleod** [comment]
JSBSim doesn't provide anything in terms of visual models of aircraft etc. So you're just switching the FDM (Flight Dynamics Model) when you enter say `737` in the Aircraft Model field. There is a whole separate mechanism to setup and change the visual model and have it hooked up in terms of control surfaces for visual movement etc. Look at the UE plug-in documentation and source code.

**Reply by MariaALJ** [reply]
Okay got it, thanks
